packages/optimizer/test/constantFolding.test.ts

```diff
@@ -0,0 +1,442 @@
+import { getRawNodes } from '../src/commons'
+import type { RawRules } from '../src/commons'
+
+import constantFolding from '../src/constantFolding'
+
+import { callWithEngine } from './utils.test'
+
+function constantFoldingWith(rawRules: RawRules): RawRules {
+	const res = callWithEngine(constantFolding, rawRules)
+	return getRawNodes(res)
+}
+
+describe('Constant folding optim', () => {
+	it('∅ -> ∅', () => {
+		expect(constantFoldingWith({})).toStrictEqual({})
+	})
+	it('Should remove empty nodes', () => {
+		expect(
+			constantFoldingWith({
+				ruleA: null,
+				ruleB: {
+					formule: '10 * 10',
+				},
+			})
+		).toStrictEqual({
+			ruleB: {
+				valeur: 100,
+				'est compressée': true,
+			},
+		})
+	})
+	it('Referenced constant should be replaced - [1 dependency]', () => {
+		const rawRules = {
+			ruleA: {
+				titre: 'Rule A',
+				formule: 'B . C * 3',
+			},
+			'ruleA . B . C': {
+				valeur: '10',
+			},
+		}
+		expect(constantFoldingWith(rawRules)).toStrictEqual({
+			ruleA: {
+				titre: 'Rule A',
+				valeur: 30,
+				'est compressée': true,
+			},
+		})
+	})
+	it('Referenced constant should be replaced - [2 dependency]', () => {
+		const rawRules = {
+			ruleA: {
+				titre: 'Rule A',
+				formule: 'B . C * D',
+			},
+			'ruleA . B . C': {
+				valeur: '10',
+			},
+			'ruleA . D': {
+				valeur: '3',
+			},
+		}
+		expect(constantFoldingWith(rawRules)).toStrictEqual({
+			ruleA: {
+				titre: 'Rule A',
+				valeur: 30,
+				'est compressée': true,
+			},
+		})
+	})
+	it('Partially compressible rule', () => {
+		const rawRules = {
+			ruleA: {
+				titre: 'Rule A',
+				formule: 'B . C * D',
+			},
+			'ruleA . D': {
+				question: "What's the value of D",
+			},
+			'ruleA . B . C': {
+				valeur: '10',
+			},
+		}
+		expect(constantFoldingWith(rawRules)).toStrictEqual({
+			ruleA: {
+				titre: 'Rule A',
+				formule: '10 * D',
+				'est compressée': true,
+			},
+			'ruleA . D': {
+				question: "What's the value of D",
+			},
+		})
+	})
+	it('Partially compressible rule with constant with mutliple dependencies', () => {
+		const rawRules = {
+			ruleA: {
+				titre: 'Rule A',
+				formule: 'B . C * D',
+			},
+			ruleB: {
+				formule: 'ruleA . B . C * 3',
+			},
+			'ruleA . D': {
+				question: "What's the value of D?",
+			},
+			'ruleA . B . C': {
+				valeur: '10',
+			},
+		}
+		expect(constantFoldingWith(rawRules)).toStrictEqual({
+			ruleA: {
+				titre: 'Rule A',
+				formule: '10 * D',
+				'est compressée': true,
+			},
+			ruleB: {
```

Comment:
Not sure why the folding keeps ruleB here. I guess it's because it holds a formula. 

Should we call constantFolding with a "root" parameter that tells which target to consider, and ignore all the rules that are not relevant ? 

In other words, should the compression be always relative to a set or targets ?